### PR TITLE
Fix horrible command handling for server side

### PR DESCRIPTION
**Pull request by draknyte1**
Made sure a null player object is handled correctly, so commands can be called Server side without NPE's being thrown. Edited via Github Web, someone should probably pull it to their IDE to check I didn't miss anything. :)

### CODE REVIEW

**Comment by Dream-Master** [comment]
@Namikon @bartimaeusnek @richardhendrics @mitchej123 @Technus can you check? thanks

File: src/main/java/com/dreammaster/command/CustomDropsCommand.java [hunk 1]
```diff
@@ -103,14 +102,17 @@ else if("reload".equalsIgnoreCase(pArgs[0]))
 
   private boolean InGame( ICommandSender pCmdSender )
   {
-      return pCmdSender instanceof EntityPlayer;
+    if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/CustomDropsCommand.java]: not needed due to instanceof

File: src/main/java/com/dreammaster/command/CustomFuelsCommand.java [hunk 2]
```diff
@@ -91,14 +90,17 @@ else if ("reload".equalsIgnoreCase(pArgs[0]))
 
     private boolean InGame(ICommandSender pCmdSender)
     {
+        if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/CustomFuelsCommand.java]: not needed aswell

File: src/main/java/com/dreammaster/command/CustomFuelsCommand.java [hunk 3]
```diff
@@ -109,6 +111,9 @@ private void SendHelpToPlayer(ICommandSender pCmdSender)
     @Override
     public boolean canCommandSenderUseCommand(ICommandSender pCommandSender)
     {
+        if (pCommandSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/CustomFuelsCommand.java]: same here

File: src/main/java/com/dreammaster/command/CustomToolTipsCommand.java [hunk 4]
```diff
@@ -91,14 +90,17 @@ else if ("reload".equalsIgnoreCase(pArgs[0]))
 
     private boolean InGame(ICommandSender pCmdSender)
     {
+        if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/CustomToolTipsCommand.java]: same here

File: src/main/java/com/dreammaster/command/CustomToolTipsCommand.java [hunk 5]
```diff
@@ -109,6 +111,9 @@ private void SendHelpToPlayer(ICommandSender pCmdSender)
     @Override
     public boolean canCommandSenderUseCommand(ICommandSender pCommandSender)
     {
+        if (pCommandSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/CustomToolTipsCommand.java]: same here

File: src/main/java/com/dreammaster/command/HazardousItemsCommand.java [hunk 6]
```diff
@@ -221,6 +220,9 @@ else if ("reload".equalsIgnoreCase(pArgs[0]))
 
     private boolean InGame(ICommandSender pCmdSender)
     {
+        if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/HazardousItemsCommand.java]: same here

File: src/main/java/com/dreammaster/command/HazardousItemsCommand.java [hunk 7]
```diff
@@ -352,6 +355,9 @@ private void ProcessAddPotionEffectCommand(EntityPlayer pPlayer, ItemStack pInHa
     @Override
     public boolean canCommandSenderUseCommand(ICommandSender pCommandSender)
     {
+        if (pCommandSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/HazardousItemsCommand.java]: same here

File: src/main/java/com/dreammaster/command/ItemInHandInfoCommand.java [hunk 8]
```diff
@@ -105,13 +104,20 @@ private String getFluidContainerContents(ItemStack pItemInQuestion)
         return tResult;
     }
     
-    private boolean InGame(ICommandSender pCmdSender)
-    {
-        return pCmdSender instanceof EntityPlayer;
+  private boolean InGame( ICommandSender pCmdSender )
+  {
+    if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/ItemInHandInfoCommand.java]: same here

File: src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java [hunk 9]
```diff
@@ -60,7 +60,11 @@ public List getCommandAliases()
   }
 
   private void moarArgs( ICommandSender pCmdSender )
-  {
+  {    
+    if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java]:
```suggestion
    if (!(pCmdSender instanceof EntityPlayer)){
```

File: src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java [hunk 10]
```diff
@@ -148,20 +152,28 @@ else if("oilstruct".equalsIgnoreCase(pArgs[0]))
     catch( Exception e )
     {
       e.printStackTrace();
-      PlayerChatHelper.SendError( pCmdSender, "Unknown error occoured [%s]", e.getMessage() );
+      if (pCmdSender != null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java]:
```suggestion
      if (pCmdSender instanceof EntityPlayer){
```

File: src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java [hunk 11]
```diff
@@ -148,20 +152,28 @@ else if("oilstruct".equalsIgnoreCase(pArgs[0]))
...
+      }
+      else {
+        MainRegistry.Logger.info("Unknown error occoured "+e.getMessage());        
+      }      
     }
   }
 
   private boolean InGame( ICommandSender pCmdSender )
   {
+    if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/AllPurposeDebugCommand.java]: same here, not needed

File: src/main/java/com/dreammaster/command/ItemInHandInfoCommand.java [hunk 12]
```diff
@@ -105,13 +104,20 @@ private String getFluidContainerContents(ItemStack pItemInQuestion)
...
+      return false;
     }
+    return pCmdSender instanceof EntityPlayer;
+}
 
     private void SendHelpToPlayer(ICommandSender pCmdSender)
     {
+        if (pCmdSender == null){
```
Comment (bartimaeusnek) [inline src/main/java/com/dreammaster/command/ItemInHandInfoCommand.java]:
```suggestion
        if(!(pCmdSender instanceof EntityPlayer)){
```

**Comment by draknyte1** [comment]
We discussed that yesterday after I’d committed this. Happy to adjust. 

I’ve never let null instance check and assumed it would throw an exception. Apparently not the case. 

*shakes fist at Java’s inconsistent null handling*

**Comment by Dream-Master** [comment]
@draknyte1 @bartimaeusnek any news here?

**Comment by richardhendricks** [comment]
I'll make a ticket to check on this.

**Comment by richardhendricks** [comment]
I removed WarpTheory from a test instance to get past the error there, and implemented the fixes above accord to Bart.  I am receiving an error like this

```
/help
[22:50:31] [Server thread/INFO]: An unknown error occurred while attempting to perform this command
[22:50:31] [Server thread/ERROR]: Couldn't process command: 'help'
java.lang.IllegalArgumentException: Comparison method violates its general contract!
        at java.util.ComparableTimSort.mergeHi(Unknown Source) ~[?:1.8.0_191]
        at java.util.ComparableTimSort.mergeAt(Unknown Source) ~[?:1.8.0_191]
        at java.util.ComparableTimSort.mergeForceCollapse(Unknown Source) ~[?:1.8.0_191]
        at java.util.ComparableTimSort.sort(Unknown Source) ~[?:1.8.0_191]
        at java.util.Arrays.sort(Unknown Source) ~[?:1.8.0_191]
        at java.util.Arrays.sort(Unknown Source) ~[?:1.8.0_191]
        at java.util.ArrayList.sort(Unknown Source) ~[?:1.8.0_191]
        at java.util.Collections.sort(Unknown Source) ~[?:1.8.0_191]
        at net.minecraft.command.CommandHelp.func_71534_d(SourceFile:84) ~[au.class:?]
        at net.minecraft.command.CommandHelp.func_71515_b(SourceFile:38) ~[au.class:?]
        at net.minecraft.command.CommandHandler.func_71556_a(CommandHandler.java:94) [z.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71333_ah(DedicatedServer.java:370) [lt.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:335) [lt.class:?]
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:427) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685) [li.class:?]
```

**Comment by richardhendricks** [comment]
this also broke /help in the client

**Comment by richardhendricks** [comment]
works fine in single player

**Comment by Namikon** [comment]
This is way too overcomplicated.
So first of all; yes, instanceof returns false if the left side is null. As such any == null is redundant.
Second: PlayerChatHelper.SendPlain() was designed to send a chat message to both Ingame and Console (As such the "plain" hint in the function name). No need to utilize the Logger.

What was the initial Error message that caused this Web-Editor commit?
I'd start from scratch with a proper error Message rather than trying to clean up this attempt

**Comment by richardhendricks** [comment]
Filed a separate PR that fixes the issue properly

**Comment by richardhendricks** [comment]
I mean, *working* on a new PR that will fix the issue properly. compareTo returning 0 is incorrect and breaking the command sort algo.

**Comment by richardhendricks** [comment]
So PlayerChatHelper will properly work if the command initiator is MinecraftServer?

**Comment by richardhendricks** [comment]
@Namikon I don't see SendPlain() in the repo at https://github.com/GTNewHorizons/GTSU/blob/master/src/main/java/eu/usrv/gtsu/helper/PlayerChatHelper.java ?

**Comment by Namikon** [comment]
Wrong repository.
PlayerChatHelper is part of YAMCore
https://github.com/GTNewHorizons/Yamcl/blob/master/src/main/java/eu/usrv/yamcore/auxiliary/PlayerChatHelper.java#L33

**Comment by Dream-Master** [comment]
why you @richardhendricks close this pr. ?

**Comment by Dream-Master** [comment]
ok

**Comment by draknyte1** [comment]
compareTo Being improperly handled by *most* commands is fixed by HelpFixer..

**Comment by richardhendricks** [comment]
Helpfixer gives me an exception when I try to run it, something about "unable to find class getMinecraftServer"

**Comment by draknyte1** [comment]
Probably a core mod so won’t work by default in dev if it’s using hardcoded class references. (Lots of core mods don’t assume their ASM will load in dev)

**Comment by richardhendricks** [comment]
I know you're speaking English, but all I hear is a loud buzzing. 🐝 

I was to try fixing these in the source repos vs just bandaiding it with HelpFixer.  But it would be nice to get a list of the offending mods...

@Dream-Master @Namikon Can we fork CraftTweaker? Unbelievably they don't tag their releases, but we can probably figure out a good place to do a custom build for us.

**Comment by draknyte1** [comment]
HelpFixer Dumps all offenders during startup logging. You’ll see every mod that adds commands and their mother are culprits. It’s one of the things that nobody ever documented and it’s kinda important. 

In saying that, I have no idea what to actually do in that function.

**Comment by Dream-Master** [comment]
sure we can fork craft tweaker because it is a fork of mine tweaker. Maybe we need to rename it. Any idea @richardhendricks

**Comment by richardhendricks** [comment]
I think we can leave the name the same, we will just have a custom version number based off our repo.

**Comment by bartimaeusnek** [comment]
@richardhendricks if you want to use core mods in a dev enviroment you need to add
-Dfml.coreMods.load=some.something.CoremodClass
as a VM Option, and even then you need to hope that they implemented a deobf version of their asm shit

**Pull request closed without merging**